File: 20180713/Search Insert Position.py
class Solution(object):
    def binarySearch(self, nums, i, j, target):
        if i >= j:
            if nums[i] < target:
                return i + 1
            else:
                return i
        else:
            mid = (i + j) / 2
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                return self.binarySearch(nums, i, mid - 1, target)
            else:
                return self.binarySearch(nums, mid + 1, j, target)
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        return self.binarySearch(nums, 0, len(nums) - 1, target)
        

File: 20180713/Search Insert Position.java
class Solution {
    public int binarySearch(int[] nums, int i, int j, int target){
        if (i >= j){
            if(nums[i] < target){
                return i + 1;
            }
            else{
                return i;
            }
        }
        else{
            int mid = (i + j) / 2;
            if (nums[mid] == target){
                return mid;
            }
            else if (nums[mid] > target){
                return binarySearch(nums, i, mid - 1, target);
            }
            else {
                return binarySearch(nums, mid + 1, j, target);
            }
        }
    }
    public int searchInsert(int[] nums, int target) {
        return binarySearch(nums, 0, nums.length - 1, target);
    }
}
